File: 278-first-bad-version/278-first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        
        if (isBadVersion(1)) {
		return 1;
	}

        
        let i = 1;
        let j = n;
        
        while(i < j){
            const m = Math.floor(i+(j-i)/2)+1;
            
            if(!isBadVersion(m)){
                 i = m;
            }else{
               j = m-1; 
            }
        }
        return j+1;
        
    };
};

File: 278-first-bad-version/278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
      int i = 1;
      int j = n;
        
        while(i < j){
            int m = i+(j-i)/2;
            if(isBadVersion(m)){
                j = m;
            }else{
                i = m+1;
            }
        }
        return i;
        
    }
}